core/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/BalancedShardsAllocator.java

```diff
@@ -757,25 +757,34 @@ private ShardAllocationDecision makeUnassignedDecision(final ShardRouting shard,
             float minWeight = Float.POSITIVE_INFINITY;
             ModelNode minNode = null;
             Decision decision = null;
-            if (canShortCircuit(throttledNodes)) {
+            final boolean explain = allocation.debugDecision();
+            if (throttledNodes.size() >= nodes.size() && explain == false) {
                 // all nodes are throttled, so we know we won't be able to allocate this round,
                 // so if we are not in explain mode, short circuit
                 return ShardAllocationDecision.no(UnassignedInfo.AllocationStatus.DECIDERS_NO, null);
             }
             /* Don't iterate over an identity hashset here the
              * iteration order is different for each run and makes testing hard */
-            final boolean explain = allocation.debugDecision();
-            Map<String, NodeExplanation> nodeExplanationMap = explain ? new HashMap<>() : null;
+            Map<String, WeightedDecision> nodeExplanationMap = explain ? new HashMap<>() : null;
             for (ModelNode node : nodes.values()) {
+                if ((throttledNodes.contains(node) // node is throttled OR
+                        || node.containsShard(shard)) // node already contains a copy of the shard
+                       && explain == false) { // not in explain mode
```

Comment:
this comment and the two above are not needed, they just literally repeat the code.